### PR TITLE
Add Ability to retrieve metadata from the server

### PR DESCRIPTION
**Pull request by Matthewaj**
- Added the ability to retrieve metadata in requests.

### CODE REVIEW

**Comment by HeavenFox** [comment]
Thanks Matt!
Some general comments:
- Our goal is to get the PR merged upstream, so we should avoid controversial changes. I wonder if the Docker and build changes are absolutely necessary.
- For the same reason, breaking backwards compatibility would be a big no-no. Looks like we can use the arity method to look up how many arguments a method take, and call it accordingly: https://ruby-doc.org/core-2.0.0/Method.html#method-i-arity
- We should maintain consistency with the official GRPC library as much as possible. This makes it easy to, in the future, move these RPC calls to be handled directly by GRPC rather than via GRPC-web. In the official implementation, the metadata is passed in _call argument (see https://grpc.io/docs/languages/ruby/basics/) Looks like it's defined here https://github.com/grpc/grpc/blob/master/src/ruby/lib/grpc/generic/active_call.rb .

**Comment by Matthewaj** [comment]
> Thanks Matt! Some general comments:
> 
> * Our goal is to get the PR merged upstream, so we should avoid controversial changes. I wonder if the Docker and build changes are absolutely necessary.
> * For the same reason, breaking backwards compatibility would be a big no-no. Looks like we can use the arity method to look up how many arguments a method take, and call it accordingly: https://ruby-doc.org/core-2.0.0/Method.html#method-i-arity
> * We should maintain consistency with the official GRPC library as much as possible. This makes it easy to, in the future, move these RPC calls to be handled directly by GRPC rather than via GRPC-web. In the official implementation, the metadata is passed in _call argument (see https://grpc.io/docs/languages/ruby/basics/) Looks like it's defined here https://github.com/grpc/grpc/blob/master/src/ruby/lib/grpc/generic/active_call.rb .

> Our goal is to get the PR merged upstream, so we should avoid controversial changes. I wonder if the Docker and build changes are absolutely necessary. 

The additional packages I added `libuvalkan1` and `libu2f-udev` were added because chrome seems to require them now. 
```
#6 26.84  google-chrome-stable depends on libvulkan1; however:
#6 26.84   Package libvulkan1 is not installed.
```
```
#6 26.84  google-chrome-stable depends on libu2f-udev; however:
#6 26.84   Package libu2f-udev is not installed.
```

The same thing happens with Ruby as well.

```
#13 3.266       The last version of bundler (>= 0) to support your Ruby & RubyGems was 2.3.26. Try installing it with `gem install bundler -v 2.3.26`
#13 3.266       bundler requires Ruby version >= 2.6.0. The current ruby version is 2.5.7.206.
```

Installing bundler `2.3.26` didn't seem to resolve the issue either, and upgrading to > 2.6.0 seemed to resolve the issue with the least friction. 

> For the same reason, breaking backwards compatibility would be a big no-no. Looks like we can use the arity method to look up how many arguments a method take, and call it accordingly: https://ruby-doc.org/core-2.0.0/Method.html#method-i-arity 
We should maintain consistency with the official GRPC library as much as possible. This makes it easy to, in the future, move these RPC calls to be handled directly by GRPC rather than via GRPC-web. In the official implementation, the metadata is passed in _call argument (see https://grpc.io/docs/languages/ruby/basics/) Looks like it's defined here https://github.com/grpc/grpc/blob/master/src/ruby/lib/grpc/generic/active_call.rb .

Thanks for the heads up. I'll start working on refactoring it to pass the call argument now.

File: Gemfile.lock [hunk 1]
```diff
@@ -8,22 +8,20 @@ PATH
...
   remote: https://rubygems.org/
   specs:
-    addressable (2.7.0)
-      public_suffix (>= 2.0.2, < 5.0)
-    apparition (0.1.0)
-      backports
-      capybara (~> 3.12, < 4)
+    addressable (2.8.1)
+      public_suffix (>= 2.0.2, < 6.0)
+    apparition (0.6.0)
```
Comment (chimano) [inline Gemfile.lock:13]: Had to update apparition. There was some interaction between the newer versions of chrome that just caused some tests to hang indefinitely.

File: spec/unit/grpc_web/server/message_serialization_spec.rb [hunk 2]
```diff
@@ -16,6 +16,7 @@
     let(:service) { TestHelloService.new }
     let(:service_method) { :SayHello }
     let(:accept) { '*/*' }
+    let(:metadata) { {} }
```
Comment (ivywit) [inline spec/unit/grpc_web/server/message_serialization_spec.rb]: Should this be getting added to the request above? It doesn't seem to be breaking but I figured I'd check as the variable isn't referenced anywhere from what I'm seeing. I might be missing if it's somewhere in a shared test though

File: spec/unit/grpc_web/server/text_coder_spec.rb [hunk 3]
```diff
@@ -18,6 +18,7 @@
     let(:service) { instance_double(TestHelloService) }
     let(:service_method) { :SayHello }
     let(:accept) { '*/*' }
+    let(:metadata) { {} }
```
Comment (ivywit) [inline spec/unit/grpc_web/server/text_coder_spec.rb]: Same as the previous question

**Review by ivywit — APPROVED** [review]
LGTM, just curious about some of the test variables

File: spec/unit/grpc_web/server/message_serialization_spec.rb (hunk 2 above)
Comment (chimano) [reply spec/unit/grpc_web/server/message_serialization_spec.rb]: Hmm good point I'm not too sure why this was added, I'll be removing it

File: spec/unit/grpc_web/server/text_coder_spec.rb (hunk 3 above)
Comment (chimano) [reply spec/unit/grpc_web/server/text_coder_spec.rb]: Removed

**Pull request merged**